smithy-model/src/main/java/software/amazon/smithy/model/shapes/NamedMembersShapeBuilder.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.shapes;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Consumer;
+import software.amazon.smithy.utils.BuilderRef;
+
+/**
+ * Builder used to create a List or Set shape.
+ * @param <B> Concrete builder type.
+ * @param <S> Shape type being created.
+ */
+abstract class NamedMembersShapeBuilder<B extends NamedMembersShapeBuilder<B, S>, S extends Shape>
+        extends AbstractShapeBuilder<B, S> {
+
+    protected final BuilderRef<Map<String, MemberShape>> members = BuilderRef.forOrderedMap();
+
+    @Override
+    public final B id(ShapeId shapeId) {
+        // If there are already any members set, update their ids to point to the new parent id.
+        for (MemberShape member : members.peek().values()) {
+            addMember(member.toBuilder().id(shapeId.withMember(member.getMemberName())).build());
+        }
+        return super.id(shapeId);
+    }
+
+    /**
+     * Replaces the members of the builder.
+     *
+     * @param members Members to add to the builder.
+     * @return Returns the builder.
+     */
+    @SuppressWarnings("unchecked")
+    public B members(Collection<MemberShape> members) {
+        clearMembers();
+        for (MemberShape member : members) {
+            addMember(member);
+        }
+        return (B) this;
+    }
+
+    /**
+     * Removes all members from the shape.
+     *
+     * @return Returns the builder.
+     */
+    @SuppressWarnings("unchecked")
+    public B clearMembers() {
+        members.clear();
+        return (B) this;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public B addMember(MemberShape member) {
```

Comment:
Missing docs.

Reply:
This was missing docs before too so no net change